### PR TITLE
tests: move cli/core to jest

### PR DESCRIPTION
**Pull request by patrickhulce**
in pursuit of #5383 

| Before | After |
| - | - |
| image | image |

only real code changes are before -> beforeAll, after -> afterAll, and moving .timeout

lots of noise from `eslint-env mocha` though :/

it also caught a bug that's not clear at all how the mocha tests were passing in CLI case where requiring in the yargs getFlags invokes process.exit (!!!!!). there's another psuedobug where we pass in 'Unknown' strings to our parseURL function which can throw

I'm not using jest's built-in expectations/coverage/anything yet to keep the diff small and easily revertable if folks don't like it.

### CODE REVIEW

**Comment by paulirish** [comment]
Also this is mad cool. Excited for this. :D

File: jest.config.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply jest.config.js]: done

File: jest.config.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  testEnvironment: 'node',
+  testMatch: [
+    '**/lighthouse-core/**/*-test.js',
+    '**/lighthouse-cli/**/*-test.js',
+  ],
+};
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply jest.config.js]: how important is this one? do we think 2s was a good feature? :)

in jest you do this via `jest.setTimeout` which requires a bootstrap file, and AFAICT cannot be done through config https://facebook.github.io/jest/docs/en/configuration.html

File: package.json [hunk 3]
```diff
@@ -31,9 +31,8 @@
     "test": "yarn lint --quiet && yarn unit && yarn type-check && yarn diff:sample-json",
     "test-extension": "cd lighthouse-extension && yarn test",
     "test-viewer": "cd lighthouse-viewer && yarn pptr-test",
-
-    "unit-core": "mocha --reporter dot \"lighthouse-core/test/**/*-test.js\"",
-    "unit-cli": "mocha --reporter dot \"lighthouse-cli/test/**/*-test.js\"",
+    "unit-core": "jest \"lighthouse-core/\"",
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply package.json]: yeah not sure how we should do this one, since we reuse lots of commands, suggestions?

[the remark this answers is not in the document]
Comment (patrickhulce) [reply package.json]: I just went with a `:ci` copy of `unit-core` 👍

File: jest.config.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulirish) [reply jest.config.js]: oic. hmmmmmmmm well nevermind then.

**Comment by paulirish** [comment]
> coverage

for whatever reason this appears to drop our coverage % down quite a bit. not sure what's happening, though. https://codecov.io/gh/GoogleChrome/lighthouse/pull/5386

**Comment by patrickhulce** [comment]
OK I sorted out the coverage problem thanks to  https://github.com/facebook/jest/issues/3190#issuecomment-354758036 😃 🎉

File: lighthouse-core/report/html/renderer/util.js [hunk 4]
```diff
@@ -288,12 +288,16 @@ class Util {
    * @return {{file: string, hostname: string, origin: string}}
    */
   static parseURL(url) {
-    const parsedUrl = new URL(url);
-    return {
-      file: Util.getURLDisplayName(parsedUrl),
-      hostname: parsedUrl.hostname,
-      origin: parsedUrl.origin,
-    };
+    try {
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/util.js]: why was this needed?

File: lighthouse-core/test/gather/computed/manifest-values-test.js [hunk 5]
```diff
@@ -28,7 +28,7 @@ function noUrlManifestParser(manifestSrc) {
   return manifestParser(manifestSrc, EXAMPLE_MANIFEST_URL, EXAMPLE_DOC_URL);
 }
 
-/* eslint-env mocha */
+/* eslint-env jest */
```
Comment (brendankenny) [inline lighthouse-core/test/gather/computed/manifest-values-test.js]: delete this one?

File: package.json [hunk 6]
```diff
@@ -31,20 +31,22 @@
...
     "core-unit": "yarn unit-core",
     "cli-unit": "yarn unit-cli",
     "viewer-unit": "yarn unit-viewer",
     "watch": "yarn unit-core --watch",
 
-    "unit:silentcoverage": "nyc --silent yarn unit && nyc report --reporter text-lcov > unit-coverage.lcov",
-    "coverage": "nyc yarn unit && nyc report --reporter html",
-    "smoke:silentcoverage": "nyc --silent yarn smoke && nyc report --reporter text-lcov > smoke-coverage.lcov",
-    "coverage:smoke": "nyc yarn smoke && nyc report --reporter html",
+    "unit:silentcoverage": "nyc --silent --clean yarn unit:ci && nyc report --reporter text-lcov > unit-coverage.lcov",
```
Comment (brendankenny) [inline package.json]: one unfortunate side of this (at least judging from that comment in the jest issue) is that subprocesses don't automatically collect coverage. I don't *believe* we do that in any unit tests, but it is a handy thing to lose

Comment (brendankenny) [reply package.json]: since jest automatically collects coverage with `--coverage`, are the lcov files written anywhere so we drop the `nyc` calls (for non-smoke tests) altogether?

File: package.json [hunk 7]
```diff
@@ -31,20 +31,22 @@
...
+    "coverage": "nyc --clean yarn unit:ci && nyc report --reporter html",
+    "smoke:silentcoverage": "nyc --silent --clean yarn smoke && nyc report --reporter text-lcov > smoke-coverage.lcov",
+    "coverage:smoke": "nyc --clean yarn smoke && nyc report --reporter html",
```
Comment (brendankenny) [inline package.json]: FWIW I believe `--clean` is the default

File: lighthouse-core/report/html/renderer/util.js (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-core/report/html/renderer/util.js]: well we only *need* this or a fix on the `_renderTextURL` to be looser on its try/catch (we check instanceof TypeError), but it seemed like a good catch on passing invalid URLs to `parseURL`. We pass through several `'Unknown'` strings to this function which throw

File: lighthouse-core/test/gather/computed/manifest-values-test.js (hunk 5 above)
Comment (patrickhulce) [reply lighthouse-core/test/gather/computed/manifest-values-test.js]: done

File: package.json (hunk 6 above)
Comment (patrickhulce) [reply package.json]: if we use jest everywhere we won't need nyc for the unit tests at all and we can use the lcov/html files they generate in `./coverage`

right now, jest coverage output is just disabled and it lets nyc collect/combine everything since we still need it for the mocha cases

File: package.json (hunk 7 above)
Comment (patrickhulce) [reply package.json]: removed

File: jest.config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (wardpeet) [reply jest.config.js]: we could go for eslint-env jest

**Comment by patrickhulce** [comment]
so what's the story here, are we sold on jest? sticking with mocha? undecided? :)

**Comment by wardpeet** [comment]
@patrickhulce I'm not hearing any people saying no that means we can go for Jest, right? 😛

File: jest.config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply jest.config.js]: ya @wardpeet I made that change 👍 :)

**Comment by patrickhulce** [comment]
I like the way you think @wardpeet, I'm thinking auto-merge in 48 hours if no more objections? ;)

**Comment by wardpeet** [comment]
They might not even notice 😄

**Comment by paulirish** [comment]
The advantages that I'm jazzed about:

* Way faster locally. 
* Display is very concise. You can see failing tests immediately, but passing tests don't scroll the view.
* Update expectations snapshot with a keystroke sounds soooooo good.

The disadvantages that I'm bummed about:

* A bit slower on travis.


----------

On the whole though i think it'll be an improved experience.

File: package.json [hunk 8]
```diff
@@ -93,6 +96,7 @@
     "gulp": "^3.9.1",
     "gulp-replace": "^0.5.4",
     "gulp-util": "^3.0.7",
+    "jest": "^23.0.1",
```
Comment (paulirish) [inline package.json]: can now use 23.2.0

Comment (patrickhulce) [reply package.json]: done

File: lighthouse-core/report/html/renderer/util.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/report/html/renderer/util.js]: hmm, but why was it needed for this PR? If it's throwing shouldn't it have caused a failure before, too? Or is the problem something else.

Comment (patrickhulce) [reply lighthouse-core/report/html/renderer/util.js]: well again it isn't strictly *needed*, but seemed like a good idea anyhow.

full details:
in jest's jsdom env, the URL class throws an error that isn't instanceof TypeError
`_renderTextURL` which normally catches TypeErrors thrown by this method was then throwing

this isn't needed anymore because along the way I turned off default jsdom environment, but it seemed like a good idea to keep, it seems like you don't think so though so I'll revert :)
